modules/indexing/src/test/java/org/apache/ignite/internal/metric/SqlStatisticsMemoryQuotaTest.java

```diff
@@ -0,0 +1,423 @@
+         */
+        private static final int PROCESS_ROWS_TO_SUSPEND = TABLE_SIZE / 2;
+
+        /**
+         * Latch to await till full scan query that uses this class function have done some job, so some memory is
+         * reserved.
+         */
+        public static volatile CountDownLatch qryIsInTheMiddle;
+
+        /**
+         * This latch is released when query should continue it's execution after stop in the middle.
+         */
+        private static volatile CountDownLatch resumeQryExec;
+
+        static {
+            refresh();
+        }
+
+        /**
+         * Refresh syncs.
+         */
+        public static void refresh() {
+            if (qryIsInTheMiddle != null) {
+                for (int i = 0; i < qryIsInTheMiddle.getCount(); i++)
+                    qryIsInTheMiddle.countDown();
+            }
+
+            if (resumeQryExec != null)
+                resumeQryExec.countDown();
+
+            qryIsInTheMiddle = new CountDownLatch(PROCESS_ROWS_TO_SUSPEND);
+
+            resumeQryExec = new CountDownLatch(1);
+        }
+
+        /**
+         * See {@link #qryIsInTheMiddle}.
+         */
+        public static void awaitQueryStopsInTheMiddle() throws InterruptedException {
+            boolean reached = qryIsInTheMiddle.await(WAIT_OP_TIMEOUT_SEC, TimeUnit.SECONDS);
+
+            if (!reached)
+                throw new IllegalStateException("Unable to wait when query starts. Test is broken.");
+        }
+
+        /**
+         * See {@link #resumeQryExec}.
+         */
+        public static void resumeQueryExecution() {
+            resumeQryExec.countDown();
+        }
+
+        /**
+         * Sql function used to suspend query when half of the table is processed. Should be used in full scan queries.
+         *
+         * @param ret number to return.
+         */
+        @QuerySqlFunction
+        public static long suspendHook(long ret) throws InterruptedException {
+            qryIsInTheMiddle.countDown();
+
+            if (qryIsInTheMiddle.getCount() == 0) {
+                boolean reached = resumeQryExec.await(WAIT_OP_TIMEOUT_SEC, TimeUnit.SECONDS);
+
+                if (!reached) {
+                    IllegalStateException exc =
+                        new IllegalStateException("Unable to wait when to continue the query. Test is broken.");
+
+                    // In some error cases exceptions from sql functions are ignored. Duplicate it in the log.
+                    log.error("Test error.", exc);
+
+                    throw exc;
+                }
+            }
+
+            return ret;
+        }
+    }
+
+    /**
+     * Functional interface to validate memory metrics values.
+     */
+    private static interface MemValidator {
+        /**
+         *
+         * @param free freeMem metric value.
+         * @param max maxMem metric value.
+         */
+        void validate(long free, long max);
+    }
+}
```

Comment:
would be nice to have a test that checks no memory quota accounting is performed when `IGNITE_DEFAULT_SQL_QUERY_MEMORY_LIMIT` is set to `-1`

Reply:
Wonderful idea! Updated!